### PR TITLE
kv/RocksDBStore: using PinnableSlice to avoid extra value copy

### PR DESCRIPTION
**Pull request by yuyuyu101**
Signed-off-by: Haomai Wang <haomai@xsky.com>

### CODE REVIEW

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -1085,31 +1085,32 @@ int RocksDBStore::get(
     const std::set<string> &keys,
     std::map<string, bufferlist> *out)
 {
+  rocksdb::PinabSlice value;
```
Comment (tchaikov) [inline src/kv/RocksDBStore.cc]:
```suggestion
  rocksdb:: PinnableSlice value;
```

see 
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/kv/RocksDBStore.cc: In member function 'virtual int RocksDBStore::get(const string&, const std::set<std::__cxx11::basic_string<char> >&, std::map<std::__cxx11::basic_string<char>, ceph::buffer::v15_2_0::list>*)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/kv/RocksDBStore.cc:1088:12: error: 'PinabSlice' is not a member of 'rocksdb'; did you mean 'PinnableSlice'?
 1088 |   rocksdb::PinabSlice value;
      |            ^~~~~~~~~~
      |            PinnableSlice
```

also i don't think this change can avoid memcpy, unless `bufferlist` is able to claim the ownership of the buffer possessed by the `PinnableSlice` instance or we can preallocate a Slice/string which can be consumed by `PinnableSlice`.

Comment (yuyuyu101) [reply src/kv/RocksDBStore.cc]: @tchaikov 
Before, rocksdb memtable/blockcache/sstfile copy to string, then append string to buffer. Two copies.

After, value refer to the string in rocksdb memtable/blockcache/sstfile, then append to buffer. one copy

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-03-20_13:25:17-rados-wip-kefu-testing-2020-03-20-1836-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-21_04:20:53-rados-wip-kefu-testing-2020-03-21-1031-distro-basic-smithi/

File: src/kv/RocksDBStore.cc (hunk 1 above)
Comment (tchaikov) [reply src/kv/RocksDBStore.cc]: @yuyuyu101 thanks for the explanation. thought you meant to say we could save the second copy. will take a closer look at rocksdb to understand how the first copy is eliminated.

File: src/kv/RocksDBStore.cc [hunk 2]
```diff
@@ -1085,31 +1085,32 @@ int RocksDBStore::get(
     const std::set<string> &keys,
     std::map<string, bufferlist> *out)
 {
+  rocksdb::PinnableSlice value;
   utime_t start = ceph_clock_now();
   auto cf = get_cf_handle(prefix);
   if (cf) {
     for (auto& key : keys) {
-      std::string value;
+      value.Reset();
```
Comment (tchaikov) [inline src/kv/RocksDBStore.cc]: just a nit, would be better if we can call Reset() after using the internal buffer. i.e., move this line after ```(*out)[key].append(value.data(), value.size());```.

Comment (yuyuyu101) [reply src/kv/RocksDBStore.cc]: done

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-03-25_08:14:01-rados-wip-kefu-testing-2020-03-25-1220-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-25_10:52:52-fs-wip-kefu-testing-2020-03-25-1220-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-25_10:54:21-rgw-wip-kefu-testing-2020-03-25-1220-distro-basic-smithi/

**Pull request merged**